lib/src/model/puzzle/history_provider.dart

```diff
@@ -0,0 +1,70 @@
+import 'dart:async';
+
+import 'package:async/async.dart';
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+import 'package:fast_immutable_collections/fast_immutable_collections.dart';
+import 'package:lichess_mobile/src/utils/riverpod.dart';
+import 'package:lichess_mobile/src/model/puzzle/puzzle.dart';
+import 'package:lichess_mobile/src/model/puzzle/puzzle_repository.dart';
+
+part 'history_provider.g.dart';
+part 'history_provider.freezed.dart';
+
+const _maxPuzzles = 500;
+
+@riverpod
+Future<IList<PuzzleHistoryEntry>> puzzleRecentActivity(
+  PuzzleRecentActivityRef ref,
+) {
+  ref.cacheFor(const Duration(minutes: 10));
+  final repo = ref.watch(puzzleRepositoryProvider);
+  return Result.release(repo.puzzleActivity(10));
+}
+
+@riverpod
+class PuzzleActivity extends _$PuzzleActivity {
+  PuzzleRepository? _repo;
+  @override
+  Future<PuzzleActivityState> build() async {
+    ref.cacheFor(const Duration(minutes: 10));
+    _repo = ref.read(puzzleRepositoryProvider);
+    final value = await ref.read(puzzleRecentActivityProvider.future);
+    return PuzzleActivityState(list: value, isLoading: false, hasMore: true);
+  }
+
+  void getNext() {
+    final currentVal = state.requireValue;
+    if (currentVal.list.length < _maxPuzzles) {
+      state = AsyncData(currentVal.copyWith(isLoading: true));
+      Result.release(
```

Comment:
Error handling is missing here.
Instead of `Result.release` you can use `fold` from `result_extensions` package to handle error:
```dart
_repo!.puzzleActivity(50, before: currentVal.list.last.date).fold(
  (value) => ....,
  (error, stackTrack) => AsyncValue.error(error, stackTrace ?? StackTrace.current),
);
```

Reply:
I wrote that too fast, maybe we should ignore the error instead, otherwise the whole list will become an error.

Reply:
We could ignore the error or show a snackbar on error

Reply:
added the snackbar